lcls-twincat-motion/Library/POUs/Motion/FB_MotionStage.TcPOU

```diff
@@ -44,8 +44,8 @@ fbDriveVirtual(En:=TRUE,
 	fPosition:=stMotionStage.fPosition,
 	fAcceleration:=stMotionStage.fAcceleration,
 	fDeceleration:=stMotionStage.fDeceleration,
-	bLimitFwd:=stMotionStage.bLimFwd,
-	bLimitBwd:=stMotionStage.bLimBwd,
+	bLimitFwd:=stMotionStage.bLimFwd AND (stMotionStage.bLimGntFwd OR NOT stMotionStage.bGantryAxis),
+	bLimitBwd:=stMotionStage.bLimBwd AND (stMotionStage.bLimGntBwd OR NOT stMotionStage.bGantryAxis),
```

Comment:
We need to rewrite this in a way that it becomes re-usable when we do the MPS axes, so the MPS FB that has custody of MC_Power doesn't need to be rewritten every time we want to adjust the virtual limit code.

I propose adding new variables to `DUT_MotionStage`, something like `bLimAllFwd` and `bLimAllBwd` that get passed into this FB as-is, then we calculate them using the truth logic you have here. Then the MPS code simply needs to pass this summary variable straight through when appropriate.

Reply:
Despite what I have written for the PMPS arch, I am no longer sure the MC_Power should reside in the PMPS code.

I think PMPS should merely be providing a set of directional enables. Otherwise we'll have to revamp FB_Drive.

Is bLimAllFwd/Bwd calculated in FB_MotionStage, or outside of it? eg. within  an EPS related block: bLimAllFwd := bLimAllFwd AND (EPS Logic);

Reply:
It seems to me that being able to version the code that protects the beamline and the code that moves the motor separately would help us quite a bit. There was talks months back about having to re-verify the mps every time we changed anything on the PLC...

I already revamped FB_DriveVirtual with this in mind.

I would do bLimAll inside the FB_MotionStage, with the assumption that any auxiliary function blocks to set gantry bits, etc. should come first.

It's also possible that things like the gantry FBs should just be embedded inside the FB_MotionStage and run or not based on the settings in the struct.

Reply:
It's a good point, keeping them totally separate. I guess the PMPS governor would just be the last thing to touch the AXIS_REF? Can we use two separate MC_Power blocks on the same axis?

What did you revamp in FB_DriveVirtual?

I think putting the gantry stuff in the stageblock is heading towards bloat.

Reply:
All I did to FB_DriveVirtual was eliminate compiler warnings and allow skipping MC_Power.

It is my understanding that using two MC_Power blocks is an error, but I'll double check.

I think would be ideal if the stage block did "everything" related to that data type, so much as these things are standard, just delegating to sub-FBs as appropriate. The block as it exists right now needs to be refactored as such.

Reply:
OK, so running two MC_Power blocks means that whichever block runs second takes effect, and the first block is ignored...

Reply:
Doesn't sound ideal.

Reply:
It's pretty bad if we are relying on MC_Power for PMPS. We need to figure out the safest way to handle this so we can compartmentalize the PMPS properly.

In the end though, the do MC_Power only in the PMPS code is probably correct, though I suspect we want PMPS to run last in the cycle anyway so it has the final say.

Reply:
Ahh now I get what you're saying about including the gantry functionality inside the block. This can be approached either way. Not sure which is best yet.

Either we use the FB_MotionStage in an overarching gantry block, or we implement all the gantry functionality within FB_MotionStage. Either way, I believe we can finish this pull request without making a decision right now.